### ISSUE TITLE
Architectural implementation of GPU-enabled NMS on raw detection output in the Segment class, exclusively for export (format='ONNX'), in ultralytics/nn/modules/head.py for YOLOv8, with the objective of minimizing latency.

### ISSUE THREAD

**Issue by chandra-ps612** [ultralytics/ultralytics#18386]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question
Hello @glenn-jocher, @RizwanMunawar, @Burhan-Q 

For minimizing latency and offloading yolov8 post-process NMS functionality entirely on GPU, I did architectural implementation of GPU enabled NMS(compatible with ONNX and TensorRT) on raw-detection output in ```Segment class``` of ```ultralytics/nn/modules/head.py```, exclusively for  ```yolov8 onnx export```
that ultimately succeeded albeit I didn't tested exported model as of now.

**Python SDK** 
```
Python 3.10.11 (main, Apr 20 2023, 19:02:41) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ultralytics import YOLO
>>> model=YOLO(" /home/ultralytics/yolov8n-seg.pt")
>>> model.export(format="onnx", device=0)
Ultralytics YOLOv8.2.9 🚀 Python-3.10.11 torch-2.1.2+cu118 CUDA:0 (NVIDIA GeForce GTX 1050 Ti, 4032MiB)
YOLOv8n-seg summary (fused): 195 layers, 3404320 parameters, 0 gradients

PyTorch: starting from ' /home/ultralytics/yolov8n-seg.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 84, 1105), (1, 32, 8400), (1, 32, 160, 160)) (6.7 MB)

ONNX: starting export with onnx 1.13.0 opset 17...
ONNX: export success ✅ 0.6s, saved as './yolov8n-seg.onnx' (13.2 MB)

Export complete (1.6s)
Results saved to /home/ultralytics
Predict:         yolo predict task=segment model= /home/ultralytics/yolov8n-seg.onnx imgsz=640  
Validate:        yolo val task=segment model= /home/ultralytics/yolov8n-seg.onnx imgsz=640 data=coco.yaml  
Visualize:       https://netron.app
' /home/ultralytics/yolov8n-seg.onnx'
```
**Apart from it, I added a new node named ```output2``` in the ```export_onnx``` function of ```exporter.py```. Please refer the onnx graph below-**
![Screenshot from 2024-12-25 10-50-51](https://github.com/user-attachments/assets/86815e6f-dcd7-4e7b-9060-46ba16e3ce20)
![Screenshot from 2024-12-25 10-50-39](https://github.com/user-attachments/assets/0d2bc4d8-d325-4573-9a7a-cc6125498737)

As per my understanding, 
when i break down ```shape``` of tensor obtained from ```x = Detect.forward(self, x)```, i decode something like this below-
1. x.shape[0] represents the batch size that is 1.
2. x.shape[1] represents the combination of 80 class confidence scores + BBoxes coordinates (x, y, w, h) that is 84.
3. x.shape[2] represents the total number of detection in an input image that is 8400

When i break down ```shape``` of ```mask_coefficient``` in onnx_graph that is represented as ```output1```, i decode something like this below- 
1. output1.shape[0] represents the batch size that is 1.
2. output1.shape[1] represents the number of mask_coefficients(mc-It modifies the protos to create the final mask for each object) that is 32.
3. output1.shape[2] __________________________________? 
**(No idea about it! Also, i have a doubt shall i have to change "the last dimension of output1" as per "the last dimension of output0" while exporting model into onnx)**

When i break down ```shape``` of ```protos``` in onnx_graph that is represented as ```output2```, i decode something like this below- 
1. output2.shape[0] represents the batch size that is 1.
2. output2.shape[1] represents the number of protos(p-learned templates that define the structure of the segmentation mask) that is 32.
3. output2.shape[2] represents the size of segmentation feature maps that is 160x160

In a nutshell, I want to break down ```last dimension of output1``` e.g. 8400, since I have applied NMS on raw-detection so i want to know its impact on ```last dimension of output1```. Shall i have change ```last dimension of output1``` according to ```last dimension of output0```??? Please refer the onnx graph for better insight.
I hope this question is crystal clear! Looking forward for your response.

### Additional

```
Dependencies I have-
ultralytics == 8.2.73                
ultralytics-thop == 2.0.0  
torch == 2.1.2+cu118            
torchaudio == 2.1.2+cu118             
torchmetrics == 0.8.0                   
torchvision == 0.16.2+cu118            
onnx == 1.13.0                   
onnx-simplifier == 0.4.33                 
onnxruntime  == 1.18.1                 
onnxsim  == 0.4.35    
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @chandra-ps612, thank you for your interest in Ultralytics 🚀! We appreciate your detailed question and effort in diving deep into the YOLOv8 architecture.

We recommend exploring the [Ultralytics Docs](https://docs.ultralytics.com) for insights into model export and usage, including [Python usage](https://docs.ultralytics.com/usage/python/) and [exporting models](https://docs.ultralytics.com/modes/export/). Many of your questions may find relevant answers there.

If this is a 🐛 Bug Report, please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) by sharing minimal code snippets, dependencies, and steps to reproduce the issue. This is essential to efficiently analyze your implementation.

For custom training ❓ Questions specific to your changes or experiments, it’s helpful to include:
- Detailed code snippets from modified files (e.g., `Segment` class and the `export_onnx` modifications).
- A thorough breakdown of the ONNX graph changes along with any observations or errors encountered during export and validation.
- Logs or screenshots from your experiments. The diagrams provided look insightful, but additional explanations regarding changes to `output1` and `output2` would aid clarity.

### Suggestions to Proceed
Below are a few steps and checks to address your concerns:
1. Verify your environment is up to date by upgrading the `ultralytics` package along with all required dependencies. Run:
   ```bash
   pip install -U ultralytics
   ```
2. Ensure compatibility with the latest [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/).
3. To debug the `output1.shape[2]`, compare it with the expected input-output pipeline dimension flow within the original and modified `Segment` class. This might help confirm if reshaping or adjustments are necessary for alignment.
   
For further insights, consider replicating this in a verified environment:
- **Colab**: <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Kaggle**: <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  
- Add your custom exports/experiments in these cloud-based setups to ensure consistent results.

### Community Support
If you'd like to collaborate further with the community or get real-time feedback:
- Join our [Discord](https://discord.com/invite/ultralytics) 🎧 for discussions.
- Explore relevant threads or create topics on our [community forum](https://community.ultralytics.com).
- Dive into related conversations on our [Subreddit](https://reddit.com/r/Ultralytics).

### Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  
If this badge is green, the latest [Ultralytics CI tests](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) for YOLO are passing, affirming the library's compatibility with validated environments.

🚀 This is an automated response, but an Ultralytics engineer will assist you further shortly. Thanks for your patience, and happy experimenting!

**Comment by Y-T-G** [comment]
You need to fix your question. You mention 3 outputs, output0, output1, output2.

There are only 2 outputs in segmentation.

**Comment by zwf159** [comment]
> You need to fix your question. You mention 3 outputs, output0, output1, output2.
> 
> There are only 2 outputs in segmentation.
Can you help me answer this question? I'm in a hurry. If I improve my segmentation model, should I focus more on the mAP or the mask metric?
When evaluating, should we focus more on the mAP of the mask part or the mask metric itself?

**Comment by chandra-ps612** [comment]
@Y-T-G 
As I mentioned above ** I've added a new node named output2 in the export_onnx function of exporter.py.**

**Comment by Y-T-G** [comment]
What's the difference between output1 and output2?

**Comment by Y-T-G** [comment]
If you want to add NMS to ONNX, you can just use `torchvision.nms` as part of the graph.

https://github.com/ultralytics/ultralytics/issues/17149#issuecomment-2437362705

**Comment by zwf159** [comment]
> If you want to add NMS to ONNX, you can just use `torchvision.ops.nms` as part of the graph.
> 
> [#17149 (comment)](https://github.com/ultralytics/ultralytics/issues/17149#issuecomment-2437362705)
help me question

**Comment by chandra-ps612** [comment]
Here output1 is for mask_coefficient that modifies protos(p) to create segmentation mask for each object.
Output2 is for protos that is learned template that define the structure of segmentation mask.

**Comment by Y-T-G** [comment]
Why do you need to separate them? It seems like you split up output0 to output0 and output1, and output1 became output2.

But I don't see why you need to do that?

**Comment by Y-T-G** [comment]
The normal output0 would have been of shape `[N, 116 (84 + 32), 8400]`. While output1 would have been `[N, 32, 160, 160]`.

**Comment by UltralyticsAssistant** [comment]
The associated pull request [#18484](https://github.com/ultralytics/ultralytics/pull/18484) has been merged, and it includes a potential fix that may resolve this issue! 🎉  

### Relevant Key Changes from the PR:
This update introduces extended export functionalities and post-processing improvements for object detection models, addressing several limitations in the workflow. Notable changes include:
- **Non-Maximum Suppression (NMS) Integration**: Exported models (TorchScript, ONNX, OpenVINO, TensorRT, TensorFlow) now include embedded NMS, providing seamless post-processing capabilities.
- **Enhanced Post-Processing Logic**: Predictions for various tasks (detection, segmentation, pose, and OBB) have been refined, including additional attribute support, like `keypoints`, for improved handling of prediction results.
- **Flexibility in Export Configurations**: Users can now toggle NMS during the export process, enabling easier deployment customization.
- **Better Metadata Handling**: Updates for improved export compatibility across formats.
- 🆙 Version upgraded to `8.3.67` for tracking this new functionality.

These changes aim to streamline the export process while improving runtime efficiency and usability for deployment scenarios.  

### Contributors ❤️
This valuable update was made possible due to contributions from [Y-T-G](https://github.com/Y-T-G), [lifafang](https://github.com/lifafang), [boorad](https://github.com/boorad), [glenn-jocher](https://github.com/glenn-jocher), [Laughing-q](https://github.com/Laughing-q), and [inisis](https://github.com/inisis). A big thank you to everyone involved! 🙌  

### How to Test the Changes:
To verify if this resolves your issue:
- Install the latest changes directly from the `main` branch:
  ```bash
  pip install git+https://github.com/ultralytics/ultralytics.git@main
  ```
- Or you can wait for the next official release on PyPI if preferred.

Let us know if this resolves the issue or if you encounter any remaining problems. Your feedback is invaluable in ensuring the robustness of these updates.  

Thanks again for bringing up this issue! 🙏 If anything persists or additional adjustments are needed, feel free to share more details.
